kafka-4428: Kafka does not exit when it receives "Address already in use" error during startup

huxihx opened the pull request:
kafka-4428: Kafka does not exit when it receives "Address already in use" error during startup

During Acceptor initialization, if "Address already in use" error is encountered, the countdown latches for all Processors have no chance to be counted down, hence Kafka server fails to exit, pending when invoking Processor.shutdown

huxihx committed:
kafka-4428: Kafka does not exit when it receives "Address already in use" error during startup
@author: huxi2b@gmail.com
During Acceptor initialization, if "Address already in use" error is encountered, the countdown latch for all Processors instsance have no chance to be counted down, hence Kafka server fails to exit, pending when invoking Processor.shutdown

huxihx commented:
@junrao @jjkoshy please review this PR. Thanks.

junrao commented:
@amethystic : Thanks for the patch. An alternative approach is to fix AbstractServerThread such that it's always shutdownable. For example, in AbstractServerThread, we could initialize shutdownLatch with CountDownLatch(0) and in startupComplete(), set shutdownLatch to CountDownLatch(1).

huxihx commented:
@junrao Agreed, but something to be considered: Firstly, since this patch only affects the code path when the socket is already bound, it's much easier to reason about the effects. Modifying AbstractServerThread might need a thorough test covering all code paths. Secondly, initializing shutdownLatch with CountDownLatch(0) does not actually keep in line with ShutdownableThread behavior, so we might need to modify ShutdownableThread too. Does this make any sense?

huxihx commented:
@junrao any feedback about the comments? Thanks.

huxihx commented:
@junrao @jjkoshy please guide how to make progress this PR? Thanks.

huxihx commented:
@junrao could you take some time to address the comments above? Thanks

huxihx commented:
@ijuma please review this PR. Thanks.

huxihx commented:
@junrao Could you help address the comments above for this PR? Seems it's been a quite while since the last response. Thank you!

huxihx commented:
@ijuma could you please review this PR since it's been for a while nobody responded it.

ijuma commented:
@amethystic, `AbstractServerThread` is only used by `Processor` and `Acceptor`, so it seems reasonable to follow Jun's suggestion I think.

huxihx committed:
Merge remote-tracking branch 'upstream/trunk' into kafka-4428_Kafka_noexit_for_port_already_use

huxihx committed:
kafka-4428 Kafka does not exit when it receives "Address already in use" error during startup
Fix code as per Jun's suggestion to reset the shutdown latch to 1 in startupComplete

huxihx commented:
@ijuma already modified the code as per Jun's suggestion. Please kindly review. Thanks.

ijuma commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -166,7 +166,7 @@ class SocketServer(val config: KafkaConfig, val metrics: Metrics, val time: Time
 
   private val startupLatch = new CountDownLatch(1)
-  private val shutdownLatch = new CountDownLatch(1)
+  private var shutdownLatch = new CountDownLatch(0)
```
This should be volatile because we await on it from multiple threads. Also, it would be good to add a comment explaining why we start it with `0` and then change it to `1`.

Finally, it would be good to have a test. Should be as simple as starting a kafka server with the same config twice and verifying that the second instance shuts down correctly.

ijuma reviewed (commented):
Thanks, left one comment.

huxi-2b committed:
Merge remote-tracking branch 'upstream/trunk' into kafka-4428_Kafka_noexit_for_port_already_use

huxihx committed:
Fix for kafka-4428: Kafka does not exit when it receives "Address already in use" error during startup
1. added a unit test case
2. added some comments

huxihx commented:
@ijuma thanks for the comments. Already added a unit test case and some comments. Please kindly review. Thanks.

ijuma commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -190,6 +193,8 @@ private[kafka] abstract class AbstractServerThread(connectionQuotas: ConnectionQ
   protected def startupComplete() = {
     startupLatch.countDown()
+    // Activate the shutdown latch by resetting it to count 1
+    shutdownLatch = new CountDownLatch(1)
```
Shouldn't this be before `startupLatch.countDown`?

huxihx replied:
I think these two latches have no impact with each other. One is for startup and one is used for shutdown. Of course, I could put it before startupLatch.countDown.

ijuma replied:
Say you invoke `KafkaServer.shutdown()` right after `KafkaServer.startup()`, there is a potential race condition (although unlikely) where `SocketServer.shutdown` is called before the latch is initialised. Right?

huxihx replied:
agreed

ijuma commented on core/src/test/scala/unit/kafka/server/ServerStartupTest.scala:
```diff
@@ -44,6 +43,47 @@ class ServerStartupTest extends ZooKeeperTestHarness {
+    props2.put("zookeeper.connect", zooKeeperConnect + zkChroot2)
+    try {
+      val server2 = TestUtils.createServer(KafkaConfig.fromProps(props2))
+      // normally these two below methods should not be invoked
```
We should have a `fail` invocation here.

ijuma commented on core/src/test/scala/unit/kafka/server/ServerStartupTest.scala:
```diff
@@ -44,6 +43,47 @@ class ServerStartupTest extends ZooKeeperTestHarness {
+      // this is expected
+    }
+    server1.shutdown()
+    CoreUtils.delete(server1.config.logDirs)
```
We can do this in a `finally` block.

ijuma commented on core/src/test/scala/unit/kafka/server/ServerStartupTest.scala:
```diff
@@ -44,6 +43,47 @@ class ServerStartupTest extends ZooKeeperTestHarness {
+    val pathExists = zkUtils.pathExists(brokerZkPath)
+    assertTrue(pathExists)
+
+    val port = zkUtils.readDataMaybeNull(brokerZkPath)._1 match {
```
You can get the port from `server1` instead of reading it from ZK. Also, why are we using a different zk chroot?

ijuma reviewed (commented):
Thanks, left a couple of comments.

huxi-2b committed:
Merge branch 'kafka-4428_Kafka_noexit_for_port_already_use' of https://github.com/amethystic/kafka into kafka-4428_Kafka_noexit_for_port_already_use

huxi-2b committed:
kafka-4428: Kafka does not exit when it receives "Address already in use"
error during startup
Addressed @ijuma's comments and fixed the code

huxihx commented:
@ijuma  Thank you for your comments. Already addressed them. Please kindly review. Appreciated.

ijuma commented:
Thanks for the PR, LGTM. I tweaked the comments and test a little before merging to trunk.
